ext/gmp/config.m4

```diff
@@ -5,11 +5,13 @@ PHP_ARG_WITH([gmp],
 
 if test "$PHP_GMP" != "no"; then
 
+  AC_CANONICAL_HOST
   MACHINE_INCLUDES=$($CC -dumpmachine)
 
   for i in $PHP_GMP /usr/local /usr; do
     test -f $i/include/gmp.h && GMP_DIR=$i && break
     test -f $i/include/$MACHINE_INCLUDES/gmp.h && GMP_DIR=$i && break
+    test -f $i/include/$host/gmp.h && GMP_DIR=$i && break
```

Comment:
I don't think this really addresses the root issue. The problem is that we're trying to look for `gmp.h` in a specific location, even though `gmp.h` is *already* on the default include path. We should instead be checking whether the compiler sees the `gmp.h` header, e.g. by using `AC_LINK_IFELSE`.

Reply:
Since we're using pkg-config for several other extensions as well, shouldn't we use it for ext/gmp too?

Reply:
GMP does not support pkg-config, unfortunately.

Reply:
I agree that it is better to let the compiler figure out if it can see the header. My autoconf-fu just wasn't strong enough to figure out how to do that right, but I'll take another look. Shouldn't it be AC_COMPILE_IFELSE though? If we're looking for header, linking is not necessary I think?

Reply:
Maybe @petk also has ideas here on how to do this nicely.

Reply:
I tried to find a way to do it in autoconf and I couldn't find the proper magic... AC_COMPILE_IFELSE and friends use current compiler settings and there seems to be no good way to see where the include comes from to determine the right GMP_DIR. I checked all other modules and most of them do the same - enumerate directories manually. So I guess we need somebody who understands autoconf better to take a look.

Reply:
Hello. Since, there don't seem to be any best practices here, and docs don't have any info on this either, I have know idea what would be a good practice here. I would simply add another check in there and not provide the `--host` override for this. So, basically there is already a list of include paths that need to be checked for PHP case, i.e.:
* The one that user can override via the --with-gmp=/path/to/include/dir
* /usr/
* /usr/include/
* /usr/include/x86_64-linux-gnu (which is basically something like `/usr/include/${host_cpu}-${host_os}`) - all these $host* and $build* variables are already generated by config.guess and config.sub scripts (that AC_CANONICAL_HOST macro requirement in the PHP_CANONICAL_HOST_TARGET).

I think, some other projects do the same. They create a list of includes related to a particular header file and loop through them.

Also this might produce a correct path for both gcc and clang:

```sh
MACHINE_INCLUDES=$($CC -dumpmachine | $SED 's/-pc-/-/')
```

And that `pc` is actually a `$host_vendor`... so in case the `sed` should be even more abstract and portable:

```sh
MACHINE_INCLUDES=$($CC -dumpmachine | $SED "s/-$host_vendor-/-/")
```

P.S: Then in such case also ldap might suffer of the similar issues when clang is used. There is also this `-dumpmachine` option used.

cc @hughmcmaster (who might know even more on this)...

Reply:
> Also this might produce a correct path for both gcc and clang:

Unfortunately no, clang adds `-unknown` on Travis. Of course can filter that, I tried that but it seems to be too brittle - there seems to be no standard and each compiler has its own twist. gcc has print-multiarch but clang does not support it. 

The advantage of --host/$host is that you can set it to the proper value (aka one that system actually uses) if autoconf messes up and it won't break stuff, afaik.

Reply:
Yes, the -print-multiarch seems to be the name of this directory, yes. How about then another `sed`? Unless this `unknown` is also output by PHP's scripts then the `$host_vendor` might be `unknown`.
 
```sh
dnl comment here
MACHINE_INCLUDES=$($CC -dumpmachine | $SED "s/-$host_vendor-/-/" | $SED 's/-unknown//')
```

Providing the `--host` option to be able to build with clang is from user point of view a bit hard to figure out. I would like to be able to build php using clang with only something like this:

```sh
./configure CC=clang
```

and let those m4 files figure out what other things are needed for clang specifics. (in ideal case scenario of course).

Reply:
Thanks for the mention, @petk. Always happy to be involved.

If you want to handle gcc and clang at the same time, you can simply do this:
```
-  MACHINE_INCLUDES=$($CC -dumpmachine)
+  MACHINE_INCLUDES=$($CC -print-multiarch 2> /dev/null || ($CC -dumpmachine | sed 's/-pc//'))
```

That said, the detection code in `ext/gmp/config.m4` needs an overhaul. `AC_CHECK_HEADERS` works well, but doesn't provide the path. I'm hoping there is a macro we can use for that, or even write one.

Reply:
> I tried to find a way to do it in autoconf and I couldn't find the proper magic... AC_COMPILE_IFELSE and friends use current compiler settings and there seems to be no good way to see where the include comes from to determine the right GMP_DIR. I checked all other modules and most of them do the same - enumerate directories manually. So I guess we need somebody who understands autoconf better to take a look.

I think there are two different cases here: Manually specified GMP_DIR and use of system libgmp. If system libgmp is used, then both the include path and the lib path should already have the GMP headers / library and actually figuring out what the directories are shouldn't be necessary.

It might be easiest to just split these cases: If a path is specified, check for gmp.h as we do now. If it itsn't then use AC_LINK_IFELSE without any extra -I or -L to ensure both the library and header is found, and add the library without an additional libpath.

Reply:
@hughmcmaster clang add -unknown, not -pc, at least on Travis system. Maybe on some other system it adds -pc too. That's the problem - it's not predictable, really. 

> It might be easiest to just split these cases

Maybe this is the best way, really.

Reply:
If this is urgent, you can filter out the `-pc` or `-unknown` part of the string with `($CC -dumpmachine | cut -d '-' -f 1,3-)`.

I've sent a patch to add pkg-config support to upstream gmp, but it's still going to take time for a new version to appear with this.